### PR TITLE
[Array API] Add linalg.diagonal

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70599

This PR adds `linalg.diagonal` following the Array API:
https://data-apis.org/array-api/latest/extensions/linear_algebra_functions.html#linalg-diagonal-x-axis1-0-axis2-1-offset-0

Fixes https://github.com/pytorch/pytorch/issues/62813

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi

Differential Revision: [D33760506](https://our.internmc.facebook.com/intern/diff/D33760506)

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -9210,6 +9212,9 @@ def ref_pairwise_distance(input1, input2):
            supports_fwgrad_bwgrad=True,
            sample_inputs_func=sample_inputs_diagonal_diag_embed),
     OpInfo('diagonal',
+           # They are not strictly aliases, but we can see them as aliases for testing purposes
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:9215]: How are they not "strictly" aliases?

**Review by mruberry — APPROVED** [review]
Cool!

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33760506).

**Comment by mruberry** [comment]
@Lezcano Internal systems can't rebase this manually - it just needs a rebase and then please ping me to merge it.

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:9215]: Becuase this one  has `-2, -1` as the default dims (good) while `torch.diagonal ` has `0, 1` as default dims (ugh).

Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:9215]: Added a comment.

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33760506).

**Pull request closed without merging**